### ISSUE TITLE
Benchmark Larger Clusters [JIRA: RIAK-1711]

### ISSUE THREAD

**Issue by rzezeski** [basho/yokozuna#44]
Yokozuna uses "doc-based" partition.  This requires it to perform a "covering" query which means the larger the cluster gets the more nodes that must be contacted.  Benchmarks should be performed for a few different cluster sizes ranging from 5-100.  Perhaps 5/10/25/50/75/100.  Scripts should be used to provision the nodes, run the tests, collect the raw results and log data, and create visualizations.  This way the tests can be run quickly and may be reproduced by others.

I think expanding the tools under `misc/bench` to deploy on Joyent's cloud is the easiest path forward.

**Comment by rzezeski** [comment]
Made a request to provision a custom riak/yokozuna image with Joyent cloud following the instructions below.  Hopefully they will fulfill my request and then I can move on to writing some bash scripts to build clusters.

http://wiki.joyent.com/wiki/display/jpc2/Creating+Your+Own+SmartMachine+Image

**Comment by rzezeski** [comment]
Last week I attempted to run a 100-node benchmark on Joyent's cloud.
A lot of things went wrong due to poor planning on my part.
- The cluster consisted of 100 "Medium 2GB (2 vCPU)" instances.
- Ring size of 1024.
- LevelDB backend, AAE on, Riak Search enabled, and of course Yokozuna
  running.
- Other than changes above pretty much default settings in
  app.config/vm.args.

I was able to start each of the 100 nodes but quickly ran into trouble
when joining.  Since I left the level buffer sizes set to the default
I was using 62G+ virtual memory, but that isn't what hurt me as none
of those pages were being written to.  The real problem involved
gossip/vnode manager, large queues, and large heaps.  When trying to
ping nodes I saw things like the following:

```
./bin/riak: line 30: whoami: not found
/home/admin/riak/rel/riak/erts-5.9.1/bin/erl: line 36: cannot fork [Not enough space]
Failed to start temp_alloc
./bin/riak: line 100: 49785: Abort(coredump)
bash: line 1: 49776 Abort
```

After a few unsuccessful attempts to get nodes joining I eventually
found a setup that allowed me to make some progress.  I was burning
$36/hr so I didn't have time to be too scientific.  I might not have
needed to do what I did but I was pressed for time.  @slfritchie was
able to stand up a 1024/5-node cluster fairly effortlessly on his
MacBook Pro.
- Switch to yessir backend while forming cluster, then to bitcask for
  benchmarks.
- Turn AAE off, disable Riak Search.
- Change ring size to 512.

While these changes did allow me to make progress it was very slow (I
even set transfer limit to 200 at one point but I think gossip
throttle was a big factor).  Given my time constraints I decided to
stop at 30 nodes sometime around 2AM.  I was able to load the fruit
data after dealing with some emfile issues on my bench runner.  On my
first query run I saw the following.

```
<html><head><title>500 Internal Server Error</title></head><body><h1>Internal Server Error</h1>The server encountered an error while processing this request:<br><pre>{error,
    {throw,
        {"Failed to search",
         "http://localhost:8093/solr/fruit/select?shards=165.225.130.29:8093/solr/fruit,165.225.131.154:8093/solr/fruit,165.225.131.231:8093/solr/fruit,165.225.131.88:8093/solr/fruit,165.225.132.216:8093/solr/fruit,165.225.133.180:8093/solr/fruit,165.225.136.146:8093/solr/fruit,165.225.137.243:8093/solr/fruit,165.225.138.1:8093/solr/fruit,165.225.138.12:8093/solr/fruit,165.225.138.13:8093/solr/fruit,165.225.138.15:8093/solr/fruit,165.225.138.17:8093/solr/fruit,165.225.138.18:8093/solr/fruit,165.225.138.19:8093/solr/fruit,165.225.138.20:8093/solr/fruit,165.225.138.22:8093/solr/fruit,165.225.138.25:8093/solr/fruit,165.225.138.26:8093/solr/fruit,165.225.138.29:8093/solr/fruit,165.225.138.30:8093/solr/fruit,165.225.138.34:8093/solr/fruit,165.225.138.35:8093/solr/fruit,165.225.138.36:8093/solr/fruit,165.225.138.37:8093/solr/fruit,165.225.138.38:8093/solr/fruit,165.225.138.39:8093/solr/fruit,165.225.138.40:8093/solr/fruit,165.225.138.5:8093/solr/fruit,165.225.138.6:8093/solr/fruit&q=apple&wt=json&fq=%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A1%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A4%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A7%29+OR+%28_yz_node%3Ariak%40165.225.138.22+AND+_yz_pn%3A10%29+OR+%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A13%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A16%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A19%29+OR+%28_yz_node%3Ariak%40165.225.138.22+AND+_yz_pn%3A22%29+OR+%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A25%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A28%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A31%29+OR+%28_yz_node%3Ariak%40165.225.138.22+AND+_yz_pn%3A34%29+OR+%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A37%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A40%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A43%29+OR+%28_yz_node%3Ariak%40165.225.138.22+AND+_yz_pn%3A46%29+OR+%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A49%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A52%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A55%29+OR+%28_yz_node%3Ariak%40165.225.138.22+AND+_yz_pn%3A58%29+OR+%28_yz_node%3Ariak%40165.225.137.243+AND+_yz_pn%3A61%29+OR+%28_yz_node%3Ariak%40165.225.138.26+AND+_yz_pn%3A64%29+OR+%28_yz_node%3Ariak%40165.225.138.25+AND+_yz_pn%3A67%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A70%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A73%29+OR+%28_yz_node%3Ariak%40165.225.136.146+AND+_yz_pn%3A76%29+OR+%28_yz_node%3Ariak%40165.225.133.180+AND+_yz_pn%3A79%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A82%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A85%29+OR+%28_yz_node%3Ariak%40165.225.136.146+AND+_yz_pn%3A88%29+OR+%28_yz_node%3Ariak%40165.225.133.180+AND+_yz_pn%3A91%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A94%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A97%29+OR+%28_yz_node%3Ariak%40165.225.136.146+AND+_yz_pn%3A100%29+OR+%28_yz_node%3Ariak%40165.225.133.180+AND+_yz_pn%3A103%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A106%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A109%29+OR+%28_yz_node%3Ariak%40165.225.136.146+AND+_yz_pn%3A112%29+OR+%28_yz_node%3Ariak%40165.225.133.180+AND+_yz_pn%3A115%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A118%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A121%29+OR+%28_yz_node%3Ariak%40165.225.136.146+AND+_yz_pn%3A124%29+OR+%28_yz_node%3Ariak%40165.225.133.180+AND+_yz_pn%3A127%29+OR+%28_yz_node%3Ariak%40165.225.138.12+AND+_yz_pn%3A130%29+OR+%28_yz_node%3Ariak%40165.225.138.1+AND+_yz_pn%3A133%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A136%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A139%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A142%29+OR+%28_yz_node%3Ariak%40165.225.138.34+AND+_yz_pn%3A145%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A148%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A151%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A154%29+OR+%28_yz_node%3Ariak%40165.225.138.34+AND+_yz_pn%3A157%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A160%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A163%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A166%29+OR+%28_yz_node%3Ariak%40165.225.138.34+AND+_yz_pn%3A169%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A172%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A175%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A178%29+OR+%28_yz_node%3Ariak%40165.225.138.34+AND+_yz_pn%3A181%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A184%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A187%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A190%29+OR+%28_yz_node%3Ariak%40165.225.138.34+AND+_yz_pn%3A193%29+OR+%28_yz_node%3Ariak%40165.225.138.30+AND+_yz_pn%3A196%29+OR+%28_yz_node%3Ariak%40165.225.138.29+AND+_yz_pn%3A199%29+OR+%28_yz_node%3Ariak%40165.225.138.35+AND+_yz_pn%3A202%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A205%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A208%29+OR+%28_yz_node%3Ariak%40165.225.138.13+AND+_yz_pn%3A211%29+OR+%28_yz_node%3Ariak%40165.225.138.18+AND+_yz_pn%3A214%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A217%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A220%29+OR+%28_yz_node%3Ariak%40165.225.138.13+AND+_yz_pn%3A223%29+OR+%28_yz_node%3Ariak%40165.225.138.18+AND+_yz_pn%3A226%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A229%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A232%29+OR+%28_yz_node%3Ariak%40165.225.138.13+AND+_yz_pn%3A235%29+OR+%28_yz_node%3Ariak%40165.225.138.18+AND+_yz_pn%3A238%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A241%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A244%29+OR+%28_yz_node%3Ariak%40165.225.138.13+AND+_yz_pn%3A247%29+OR+%28_yz_node%3Ariak%40165.225.138.18+AND+_yz_pn%3A250%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A253%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A256%29+OR+%28_yz_node%3Ariak%40165.225.138.13+AND+_yz_pn%3A259%29+OR+%28_yz_node%3Ariak%40165.225.138.18+AND+_yz_pn%3A262%29+OR+%28_yz_node%3Ariak%40165.225.138.17+AND+_yz_pn%3A265%29+OR+%28_yz_node%3Ariak%40165.225.138.15+AND+_yz_pn%3A268%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A271%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A274%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A277%29+OR+%28_yz_node%3Ariak%40165.225.138.37+AND+_yz_pn%3A280%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A283%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A286%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A289%29+OR+%28_yz_node%3Ariak%40165.225.138.37+AND+_yz_pn%3A292%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A295%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A298%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A301%29+OR+%28_yz_node%3Ariak%40165.225.138.37+AND+_yz_pn%3A304%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A307%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A310%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A313%29+OR+%28_yz_node%3Ariak%40165.225.138.37+AND+_yz_pn%3A316%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A319%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A322%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A325%29+OR+%28_yz_node%3Ariak%40165.225.138.37+AND+_yz_pn%3A328%29+OR+%28_yz_node%3Ariak%40165.225.138.36+AND+_yz_pn%3A331%29+OR+%28_yz_node%3Ariak%40165.225.138.39+AND+_yz_pn%3A334%29+OR+%28_yz_node%3Ariak%40165.225.138.38+AND+_yz_pn%3A337%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A340%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A343%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A346%29+OR+%28_yz_node%3Ariak%40165.225.138.40+AND+_yz_pn%3A349%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A352%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A355%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A358%29+OR+%28_yz_node%3Ariak%40165.225.138.40+AND+_yz_pn%3A361%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A364%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A367%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A370%29+OR+%28_yz_node%3Ariak%40165.225.138.40+AND+_yz_pn%3A373%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A376%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A379%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A382%29+OR+%28_yz_node%3Ariak%40165.225.138.40+AND+_yz_pn%3A385%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A388%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A391%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A394%29+OR+%28_yz_node%3Ariak%40165.225.138.40+AND+_yz_pn%3A397%29+OR+%28_yz_node%3Ariak%40165.225.138.20+AND+_yz_pn%3A400%29+OR+%28_yz_node%3Ariak%40165.225.138.19+AND+_yz_pn%3A403%29+OR+%28_yz_node%3Ariak%40165.225.138.5+AND+_yz_pn%3A406%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A409%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A412%29+OR+%28_yz_node%3Ariak%40165.225.138.6+AND+_yz_pn%3A415%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A418%29+OR+%28_yz_node%3Ariak%40165.225.130.29+AND+_yz_pn%3A421%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A424%29+OR+%28_yz_node%3Ariak%40165.225.138.6+AND+_yz_pn%3A427%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A430%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A433%29+OR+%28_yz_node%3Ariak%40165.225.130.29+AND+_yz_pn%3A436%29+OR+%28_yz_node%3Ariak%40165.225.138.6+AND+_yz_pn%3A439%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A442%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A445%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A448%29+OR+%28_yz_node%3Ariak%40165.225.138.6+AND+_yz_pn%3A451%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A454%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A457%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A460%29+OR+%28_yz_node%3Ariak%40165.225.138.6+AND+_yz_pn%3A463%29+OR+%28_yz_node%3Ariak%40165.225.130.29+AND+_yz_pn%3A466%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A469%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A472%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A475%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A478%29+OR+%28_yz_node%3Ariak%40165.225.130.29+AND+_yz_pn%3A481%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A484%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A487%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A490%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A493%29+OR+%28_yz_node%3Ariak%40165.225.130.29+AND+_yz_pn%3A496%29+OR+%28_yz_node%3Ariak%40165.225.131.88+AND+_yz_pn%3A499%29+OR+%28_yz_node%3Ariak%40165.225.131.154+AND+_yz_pn%3A502%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A505%29+OR+%28_yz_node%3Ariak%40165.225.131.231+AND+_yz_pn%3A508%29+OR+%28_yz_node%3Ariak%40165.225.132.216+AND+_yz_pn%3A510+AND+%28_yz_fpn%3A510+OR+_yz_fpn%3A509%29%29",
         {ok,"413",[{"Content-Length","0"},{"Connection","close"}],<<>>}},
        [{yz_solr,search,4,[{file,"src/yz_solr.erl"},{line,218}]},
         {yz_wm_search,search,2,[{file,"src/yz_wm_search.erl"},{line,69}]},
         {webmachine_resource,resource_call,3,
             [{file,"src/webmachine_resource.erl"},{line,183}]},
         {webmachine_resource,do,3,
             [{file,"src/webmachine_resource.erl"},{line,141}]},
         {webmachine_decision_core,resource_call,1,
             [{file,"src/webmachine_decision_core.erl"},{line,48}]},
         {webmachine_decision_core,decision,1,
             [{file,"src/webmachine_decision_core.erl"},{line,546}]},
         {webmachine_decision_core,handle_request,2,
             [{file,"src/webmachine_decision_core.erl"},{line,33}]},
         {webmachine_mochiweb,loop,1,
             [{file,"src/webmachine_mochiweb.erl"},{line,66}]}]}}</pre><P><HR><ADDRESS>mochiweb+webmachine web server</ADDRESS></body></html>
```

With this many partitions the filter query (`fq`) param becomes too
long and causes a `413` status from Solr.  I hot-patched this by using
a [POST to query](https://github.com/rzezeski/yokozuna/compare/master...post-get) instead.  However, looking at the size of the
filter query I also start to worry about reaching the
`maxBooleanClauses` threshold.  It can be raised but at what cost?

The other thing to notice about this reply is coverage is including
ALL 30 nodes.  Currently the coverage code optimizies to reduce vnode
count but @kellymclaughlin has some code that will allow to optimize
for minimum nodes.  In the case of Yokozuna, since there are no
vnodes, you want to minimize the number of nodes to reduce
communication overhead.

The number for the query benchmarks weren't too great considering a
30-node cluster.  Around 165 ops/s and ~315ms median latency.
However, there are some things to keep in mind here:
- There was 5-10% CPU usage across the board and 50-100% busy disk on
  many nodes (I learned of these numbers after tearing down so I was
  not able to investigate).
- For the same query Riak Search would have run only a handful of
  operations a second, if at all.  So it's still better than our
  current best.
- The numbers were consistent.

When looking at `prstat 1` output while running this benchmark I
noticed that the beam process was consistenctly using 5-10% CPU and
the Java process was using none.  Looking at the etop output I noticed
a lot of time being spent in coverage calculation.  I applied a
[cache branch](https://github.com/rzezeski/yokozuna/compare/master...rz-cache-plan-rebase) I have and that seemed to drop the beam usage to
almost nothing.  The java process stayed at 0% CPU usage as well so it
seemed as if nothing was doing any CPU work which makes me think a lot
of time was being spent waiting on network IO but I didn't have time
to verify. It's worth noting that the output that my benchmark scripts
collected from prstat don't show this drop in CPU but I remember very
distintly seeing the 0% usage (and it's in my notes frm that night) so
perhaps some files got mixed up or there is a bug in the collectors.
I dunno, at this point it was past 4AM and I had been working for
almost 18 hours.  It needs more investigation.

This was my final entry in my notes from that night which I think sums
things up well.

> It's worth keeping in mind that while the absolute perf numbers kinda
> suck the cluster was not under duress at all.  If anything you would
> think nothing was happening.  I'm not sure if riak search range
> queries or 2i would fare so well in the same situation.
> 
> I think there are 2 keys things to look into:
> 1. Where does solr dist search spend it's time
> 2. Can yokozuna be sized differently than riak_kv to reduce the amount
>    of spread (e.g. use a different ring for yokozuna?)

**Comment by rzezeski** [comment]
I'm moving this issue to the 0.5 milestone as it requires more time.

**Comment by rzezeski** [comment]
During March I ran some benchmarks against an 8-node cluster on
[baremetalcloud](http:///www.baremetalcloud.com).  The detailed report can be found [here](http://www.progski.net/yz-bm-summary/)
but a summary follows.
- Yokozuna (YZ) is more consistent than Riak Search (RS).  It holds a
  fairly flat throughput and latency profile versus RS which has great
  discrepancy between it's peaks and valleys.
- In the absolute best case for RS it beat YZ by 2.74x.  In this case
  RS only has to contact 1 node and makes use of the `fl=id`
  optimization.  YZ, on the other hand, is contacting all 8 nodes and
  doesn't have the `fl=id` optimization.
- For cases where a single term query matches more than 1000 documents
  or there is a conjunction of small and large queries RS quickly falls
  apart.  In the 'alpha' query YZ beats RS by 584x (and that's
  rounding up the RS results, it's probably closer to 700+, not only
  that but RS has errors during this query, YZ doesn't).  In the
  'papa' query YZ does 13.7x of RS.
- The story quickly deteriorates for RS if not making use of the
  `fl=id` optimization.  In that case it beats YZ only in the 'lima'
  query.  Only by 1.6x at that.
- When running concurrent 'get' + 'query' load YZ and RS come to a
  draw.  RS holds 'get' throughput better, YZ holds 'query' throughput
  better.  For most benchmark loads RS holds close to 90% of the 'get'
  throughput, YZ holds closer to 70%.  For the 'query' throughput RS
  holds around 50% whereas YZ holds more like 65%.  Once again, YZ is
  more consistent here.  Also, the `fl=id` optimization is in play.
  If it wasn't I would expect RS to perform much worse.  Finally, I
  think the reason YZ has a greater impact on 'get' throughput is
  because it involves all 8 nodes for all queries.  Changing coverage
  to optimize for nodes might help this.
- RS is the best at 2i queries.  I.e. "return all keys that match a
  given query."  All 3 solutions, 2i, RS, and YZ suck at returning
  result sets greater than 1k.  Below that RS is the clear winner
  achieving ~2k ops/s.  YZ was able to get ~1k ops/s but tails off
  quickly.  The best 2i could do was ~700 ops/s (although it did
  better than YZ for 100 and 1k queries).  Two things that would help
  Yokozuna here.  First, the default max heap size of 512MB is not
  enough.  Second, it was doing scoring for these queries and it
  doesn't need to so a lot of time could be saved by removing scoring.
- Finally, there two easy changes to be made to Yokozuna to help query
  throughput.
1. Compress the filter query (fq) param so less work needs to be done
   per query.
2. Optimize the coverage code for node usage rather than vnodes.  For
   all benchmarks Yokozuna was querying all 8 nodes for every query.
   It doesn't need to do that.  It could have queries 3 of the nodes
   and still achieved the correct result.

Work has already started on for point 1: #76

**Comment by rzezeski** [comment]
few more important things I forgot to mention in the last comment.

No matter what, as the number of nodes in a distributed search
increases the greater the latency will become.  The real question is,
what does the function range of `latency` and `# of nodes` look like?
How many nodes does it take before things become too slow?  This will
of course depend on the definition of "too slow" and the hardware
being used.  But having a general idea of how distributed search
degrades is important.

Most users, according to data collected thus far, run clusters with
less than 10 nodes.  Furthermore, what works for 10 nodes will often
not work for 100 nodes.  An order of mag can change everything.  The
point is to stay focused on the majority use-case and don't get lost
in the 100-node saga.

While most user clusters are < 10 nodes it would be very prudent to
make sure Yokozuna works well with 15-30 node clusters.  If we get the
coverage code optimization in a 30 node cluster should require 10
nodes per query.  But for a lot of indexes the corpus may be small
enough to fit on a single node.  Solr is very fast in single node
format.  We should investigate methods for which Yokozuna can be
partitioned differently from KV.  E.g. you could attach a "partition
factor" to an index that describes how many nodes that index should be
partitioned across.  Many indexes could be held in one partition and
thus all queries would be single node and very very fast.  Or you
could partition across 3 nodes and would have to query at most 3 nodes
(indexes would still be replicated so depending on how things are
implemented you don't have to hit all nodes).  Doing something like
this would complicate AAE but it should still be possible.  You could
keep the original KV partition numbers and when building the AAE trees
just make sure to query all the Yokozuna index partitions.  I think
the more complicated part is dealing with non-local writes.  This
makes index-time more akin to Riak Search thus increasing latency and
making error cases more awkward.

**Comment by rzezeski** [comment]
This is not a 2.0-RC blocker.  I know about plenty of perf issues to fix that would benefit any size cluster before running larger tests.  Pushing to 2.0.1.

**Comment by DSomogyi** [comment]
Comment for Jira.

**Comment by Basho-JIRA** [comment]
[~seancribbs], not sure what to do with this - we could throw this to [~bmcclain] to see what he thinks or just close it out.  Thoughts? 

**_[posted via JIRA by Derek Somogyi]**_

**Comment by Basho-JIRA** [comment]
I don't think there's a specific task to do here. We should be benchmarking YZ for each release, but this was probably created before we released 2.0. Closing. 

**_[posted via JIRA by Sean Cribbs]**_
